### ISSUE TITLE
"variable is None" does not work with Union

### ISSUE THREAD

**Issue by ostcar** [python/mypy#3845]
Python version: 3.6.2
Mypy version: 0.521

When a variable has the type Union with two different types **and** None,  ```varialbe is None```  does not work:

```python
def foo(a: Union[int, str, None]) -> Optional[int]:
    if a is None:
        return a
    elif isinstance(a, str):
        return 0
    return a
``` 
mypy returns for the third line:
```
 Incompatible return value type (got "Union[int, str, None]", expected "Optional[int]")
```

As a workarount the following code works:
```python
def foo(a: Union[int, str, None]) -> Optional[int]:
    if isinstance(a, type(None)):
        return a
    elif isinstance(a, str):
        return 0
    return a
```

**Comment by ilevkivskyi** [comment]
Thanks for reporting! You should use the ``--strict-optional`` flag, then everything works as expected (just tried on master). In general, it is a good idea to try this flag whenever you work with ``Optioanal``. This flag will be turned on by default at some point. I will keep this issue open, since the behaviour is counter-intuitive here (one would expect more errors with ``--strict-optional``, not less :-)

**Comment by ostcar** [comment]
Thanks, this helped.

**Comment by ilevkivskyi** [comment]
This works correctly now.
